### DISCUSSION TITLE
Ideas for QSIPrep 1.0

### DISCUSSION THREAD

**Discussion by mattcieslak** [PennLINC/qsiprep#562] [Announcements]
Hi Everyone! QSIPrep has been around for a few years now and I think there are some areas where we can make some significant improvements. Some of these have been planned for awhile and many have open issues already. I'd like to get some input from you all on the big changes though. Many of these ideas have already been implemented but feedback is still appreciated!

## Anatomical processing
 1. Allow users to pick an "anatomical contrast" of either "T1w", "T2w" or "none". "none" would replace --dwi-only and the others would select which would be used to estimate an anatomical reference. T2w is very useful for infant data. (already implemented in #553)
 2. Use synthstrip and synthseg for brain extraction and tissue segmentation. This option makes the T2w option above workable. Testing so far suggests these are considerably faster and more flexible than antsBrainExtraction and have a much better license than FAST. (in #553)
 3. Allow the user to specify a custom reference image. It would look something like --template /path/to/my/T[12]w.nii.gz where there would also need to be a brain mask in the same space. Then the rigid alignment part of qsiprep would register to this image instead of a template (which would normally produce the AC-PC alignment). This would allow users to pass in the preprocessed T1w image from fmriprep as the template and result in processed dwi data aligned to the processed bold (idea from @jyeatman). It also opens up the possibility of getting derivatives data directly on surfaces prepared my fmriprep.
 4. Change `space-T1w` to `space-T1wACPC` in the derivatives. I'd love feedback on this point because I'm waffling. `space-T1w` is consistent with fmriprep and implies that the brain shape hasn't been changed. But the outputs of qsiprep have  `space-T1w` images that aren't aligned to the actual T1w(!) and sometimes is impossible to directly transform to the T1w due to header deobliquing. I think `space-T1w` makes the most sense if the T1w from fmriprep is used as the reference. There has to be some labeling of the ACPC alignment in the bids though. @tsalo, any thoughts here? (partially implemented in #553)
 5. I'm not sure what to do about nonlinear anatomical registration. Estimating a nonlinear transform to a template space isn't *really* part of dwi preprocessing, but I like doing it during preprocessing so you can run multiple reconstructions without re-estimating the transform in each reconstruction workflow. I can also see how it could be useful to estimate this transform in the reconstruction workflow.

**Comment by mattcieslak** [comment]
## dMRI Preprocessing
1. I'd like to remove the SyN-SDC and replace it with Synb0-disco (#546). The SyN-SDC implementation is from a very old fmriprep and performance has been pretty variable. 
2. Moving some preprocessing options to the very end of the pipeline. Specifically, b1 bias correction has been moved to after the data is resampled #527 (unless you specify `--b1-biascorrect-stage legacy` or disable it with `--b1-biascorrect-stage none`). 
3. I'd also like to add an option to run `patch2self` at the very end of the pipeline (@ShreyasFadnavis). I think this is a viable method to perform outlier replacement outside of Eddy. Partially implemented in #527
4. `Gibbs` has been added from TORTOISE. #490. It can be used instead of `mrdegibbs` and is optimized for PartialFourier acquisitions #490. I'd love to see some benchmarking on this or get some feedback if anyone tries it out. Try it out with `--unringing-method rpg`
5. There is a SHORELine version for single shell acquisitions now! @nseider implemented a tensor version that can be accessed via `--hmc-model tensor` #519. I don't have CI tests for this yet, but do have anecdotal evidence it works.
6. I added the CNR-per-shell estimates to the ImageQC.csv file #541, but there are a lot of other interesting QC metrics we can estimate from the preprocessing stages. Planned additions are the FWHM of the residuals images (idea from @mharms). Very interested to hear other ideas here.

**Reply by ShreyasFadnavis** [reply]
@mattcieslak - I like the idea of doing Patch2Self at the very end. It would be ideal if we could run it before the interpolation step :) Not sure if that is feasible. We can also extract SNR and CNR measures like Eddy after P2S -- should be straightforward to implement!

I can also add it inside P2S in DIPY.

**Reply by mattcieslak** [reply]
Those extra SNR/CNR measures would be great to have! It's always possible to run it before interpolation, but I prefer to have motion correction run in a program dedicated to estimating motion. I know patch2self can end up interacting with head motion so I'm thinking the final interpolated data may be the best option.

**Comment by mattcieslak** [comment]
## dMRI Reconstruction

1. The FAST-based ACT workflows should be removed. These segmentations are not good.
2. In general I'd like to focus on tractometry/bundle analysis reconstruction workflows instead of braingraphs going forward. To that end:
3. Add a workflow for DSI Studio AutoTrack. I love this method and want to make it extremely easy to use via qsiprep @Frankyeh.
4. @36000, @arokem one of the things that AutoTrack doesn't do is along-bundle metrics (or maybe I missed them?). Are there any plans for PyAFQ to calculate these on bundles generated by AutoTrack?
5. Does anyone know of an infant-optimized bundle atlas? The adult atlas seems to work well with AutoTrack in infants and I know there's an Infant-optimized AFQ, but I'm not sure if there's a community-consensus set of bundles for this age range.
6. Mapping dMRI scalars to surfaces. There are a couple measures that are interpretable in gray matter. It would be pretty cool to get these measures into a surface format that is usable by BOLD people @gshafiei.

**Reply by arokem** [reply]
> [@36000](https://github.com/36000), [@arokem](https://github.com/arokem) one of the things that AutoTrack doesn't do is along-bundle metrics (or maybe I missed them?). Are there any plans for PyAFQ to calculate these on bundles generated by AutoTrack?

I've never tried this, but, assuming these are bundles of streamlines, it should be straightforward, as demonstrated in [this DIPY example](https://dipy.org/documentation/1.7.0/examples_built/17_streamline_analysis/afq_tract_profiles/#sphx-glr-examples-built-17-streamline-analysis-afq-tract-profiles-py)

**Reply by 36000** [reply]
One approach here is to have a way to input segmented tractographies to pyAFQ. This is a little odd because at this point you are skipping most of the pyAFQ pipeline and just doing profile extraction / visualization. However it might be easier if people are used to the pyAFQ interface. The other approach is what @arokem mentioned, which is to use the methods in dipy to extract profiles without using the pyAFQ interface. I will volunteer to implement whichever of the approaches we choose

**Reply by mrribbits** [reply]
+1 for DSI Studio workflow.  Would be nice to be able to customize the DSI Studio workflow with a json similar to eddy config since there are options one can set when creating the SRC, FIB, and TT files.

**Reply by frankyeh** [reply]
I will be happy to modify DSI Studio to help improving the pipeline. Any specific suggestions?

**Reply by mrribbits** [reply]
@frankyeh Not sure if this is possible, but can a user specify the DSI Studio command-line parameters within QSIPrep?  Like for FSL eddy, a user can create a json file with the parameters in QSIPrep.  Something similar possible with DSI Studio?  

E.g.: 
For SRC: --up_sampling
For FIB: --method --param0 --check_btable --align_acpc --mask --other_output record_odf
For TT: --fiber_count --turning_angle --step_size --min_length --max_length --method --smoothing --tip_iteraion --random_seed --roa --ter --end --export

* I suspect align_acpc should always be OFF since QSIPrep does acpc alignment already for the user? There may be other parameters that shouldn't be changed to avoid conflicting with QSIPrep, not sure.

**Reply by mattcieslak** [reply]
@mrribbits you can access most of these through the recon spec. The default values are [here](https://github.com/PennLINC/qsiprep/blob/master/qsiprep/data/pipelines/dsi_studio_gqi.json) but this file can be edited and the edited version passed to `--recon-spec`.

I'm currently working on adding the most recent version of DSI Studio and the arguments have changed quite a bit

**Reply by mrribbits** [reply]
@mattcieslak Ah that's good to know, thanks!

**Reply by mattcieslak** [reply]
also, qsiprep results will ac-pc orientation already

**Reply by smeisler** [reply]
👍 to these proposed points, few notes below:

> Does anyone know of an infant-optimized bundle atlas? The adult atlas seems to work well with AutoTrack in infants and I know there's an Infant-optimized AFQ, but I'm not sure if there's a community-consensus set of bundles for this age range.

Maybe https://www.nitrc.org/projects/uncebds_neodti / https://pubmed.ncbi.nlm.nih.gov/35401073/?

> The FAST-based ACT workflows should be removed. These segmentations are not good.

I agree they're not good, but would FAST still be a viable alternative for tough clinical cases where FreeSurfer fails?

**Comment by mattcieslak** [comment]
## Distribution and Variants

Many groups have been unable to use qsiprep due to the FSL license. A lot of work (credit goes to @jbh1091) went into making most qsiprep workflows runnable in a container that has no FSL executables or data in it. The plan is for each qsiprep release to also release a `qsiprep:${version}-nofsl` image. Users of this image will have to use `--hmc-model 3dSHORE|tensor` and `--pepolar-method DRBUDDI`. Feedback would be very welcome here!

**Reply by smeisler** [reply]
Could TORTOISE's `DIFFPREP` also be used for HMC in the `-nofsl` distribution?

**Reply by mattcieslak** [reply]
I'd love to add DIFFPREP's motion correction. It uses a very similar approach to SHORELine, but also does eddy current correction. @eurotomania would need to add some infrastructure to TORTOISE so it could be accessed on its own and could work on non-shelled acquisitions. But this is the dream

**Reply by araikes** [reply]
What about DIPY's motion correction for FSL-less methods?

**Reply by mattcieslak** [reply]
I'm not familiar with them, but am pro-DIPY in general

**Reply by araikes** [reply]
https://dipy.org/documentation/1.7.0/interfaces/motion_flow/

You can also, obviously, do it outside the CLI and get the rotated bvecs. I'm using it for ex-vivo rodent data with pretty good success.

**Reply by mattcieslak** [reply]
very interesting, it looks like they're using the b=0 as a reference image and registering all the b>0 images to that. I'm not sure how well that would work with high b-values and lots of artifacts.

**Reply by arokem** [reply]
Totally agree that this would pose some problems where b=0 and b>0 look very different. We do also have this WIP: https://github.com/nipreps/eddymotion, which uses a variety of DIPY models for prediction, and is essentially a generalization of ShoreLine (and has no FSL dependency).

**Reply by araikes** [reply]
@arokem 

I have some data for which this would be very useful.... Is there a tutorial for using?

**Reply by arokem** [reply]
Hey - sorry - no tutorial yet. You might be able to sort through this by reading through the code (e.g., this test: https://github.com/nipreps/eddymotion/blob/main/test/test_integration.py#L33), but as I said this is still WIP, and there are still some kinks to iron out.
